tools/api-client/python/lib/bmutils.py

```diff
@@ -98,10 +100,40 @@ def wrap_create_game(self, pbutton, obutton='', opponent=''):
     return retval.data
 
   def wrap_load_game_data(self, game):
-    retval = self.load_game_data(game)
-    if not retval.status == 'ok':
-      raise ValueError, "Failed to call loadGameData, got: " + retval
-    data = retval.data
+    # if we're using a cache directory
+    if self.cachedir:
+      # if the cache directory doesn't exist, create it
+      if not os.path.isdir(self.cachedir):
+        os.makedirs(self.cachedir)
+      # set the path to the cache file for this game
+      cachefile = self.cachedir + "/" + unicode(game) + ".json"
+      # if the cache already has a file for this game
+      if os.path.isfile(cachefile):
+        # get the game data from the cache
+        with open(cachefile) as cache_fh:
+          data = json.load(cache_fh)
+      # otherwise (the cache didn't already have a file for this game)
+      else:
+        # load the game
+        retval = self.load_game_data(game)
+        # if that didn't work, raise an exception
+        if not retval.status == 'ok':
+          raise ValueError, "Failed to call loadGameData, got: " + retval
+        # if we're still here, we have the game data
+        data = retval.data
+        # if the game is completed
+        if data['gameState'] == "END_GAME":
+          # put the data for this game into the cache
+          with open(cachefile, 'w') as cache_fh:
+            json.dump(data, cache_fh, indent=1, sort_keys=True)
+    # otherwise (we aren't using a cache directory), load the game
+    else:
+      retval = self.load_game_data(game)
+      if not retval.status == 'ok':
+        raise ValueError, "Failed to call loadGameData, got: " + retval
+      data = retval.data
```

Comment:
Here's a suggestion for how to structure the logic of `wrap_load_game_data()` differently, to get rid of the repetition of the code block which actually gets the game:
```python
  data = None
  if self.cachedir:
    # Do all of your cache setup stuff and try to load the game from the cache file if it exists
  if data is None:
    # Actually load the data
    if self.cachedir:
      # do the cache write stuff
```

What do you think?  The repeated block is pretty short, so it's not too bad as is, but this seems like a potentially easy way to get around it.